py/htm/optimization/README.md

```diff
@@ -0,0 +1,150 @@
+of a model for given task/dataset. 
+In *HTM.core* it is provided in python by `python -m htm.optimization.ae `. 
+
+There are several methods implemented:
+- particle swarm optimization (PSO) by `swarming.py`
+- manual search
+- exhaustive search
+
+## Requirements
+
+- Python 3
+- `ExperimentModule.py` is your python script containing the model to be optimized as
+   well as code to evaluate the models performance.
+- `ExperimentModule.default_parameters = {}`
+   Global dictionary containing all of the parameters to modify.
+   Parameters must be one of the following types: dict, tuple, float, int.
+   Parameters can be nested in multiple levels of dictionaries and tuples.
+   For more details see `htm.optimization.parameter_set`
+- `ExperimentModule.main(parameters=default_parameters, argv=None, verbose=True)`
+   Returns (float) performance of parameters, to be maximized.
+   For example, see file: `py/htm/examples/mnist.py`
+
+## Optimize your model, parameter tuning
+
+Run your experiment with the AE program:
+`$ python3 -m htm.optimization.ae [ae-arguments] ExperimentModule.py [experiment-arguments]`
+
+### Examples
+
+`python -m htm.optimization.ae -n 7 --memory_limit 14 -v --swarming 100  mnist.py`
+
+Uses PSO ("swarming") with 100 particles, 7 CPU threads, upto 14GB RAM, to optimize MNIST results in mnist.py.
+
+`python -m htm.optimization.ae --grid_search "" -n 7 --memory_limit 14 mnist.py` 
+
+Does full exhaustive search for all parameters (the `""`) on `mnist.py` script.
+
+`python -m htm.optimization.ae --grid_search '["potentialPct"]' -n 7 --memory_limit 14 mnist.py`
+
+Tries several different values for the "potentialPct" parameter.
+
+### Manual search
+
+It is possible to manually specify parameters to evaluate.  While the AE program
+is **not** running, append the following to the lab-report file:
+
+```
+================================================================================
+Modification: ['potentialPct'] = 0.66
+Modification: ['columnDimensions'][1] = 123
+```
+
+The 80 equal signs are significant!  The AE program uses them as section dividers.
+
+Each "modification" line is a snippet of python code which is applied to the
+default parameters.
+
+Then run the command: `python -m htm.optimization.ae --parse mnist.py`
+which will read the lab-report file, fill in any missing fields, and write it
+back to the same file, without running any experiments.  After running the AE
+program, the section which we appended to the lab-report will look like:
+
+```
+================================================================================
+Modification: ['columnDimensions'][1] = 123
+Modification: ['potentialPct'] = 0.66
+Hash: FFD67F4E
+Journal: mnist_ae/FFD67F4E.journal
+Attempts: 0
+Notes: 
+```
+
+Notice the "Hash" has been computed for this set of parameters.  The Hash is a
+unique identifier for these parameters.  Use the following command to evaluate
+these parameters:
```

Comment:
I'm not sure I understand what are "these parameters"? The 2 I manually set, and the rest? Will be tried by AE? Or is it the current state at the time I stopped the running AE (to fill in the `Modification:`s)?

Reply:
"these parameters" refers to the new set of parameters which you manually entered.

Also, this example adds a single new set of parameters, not 2 sets.  There are two modifications to the new parameter set.  If you want to add 2 new sets of parameters, then you need another section divider to split the modifications into two sections.

Reply:
I'm trying to figure if I understand... so the "manual" is not really a search method, it allows me to model an experiment setup and run (&record) it. (The other, not specified params, are the constructors defaults?) 
What is a usecase for this? I'd imagine I have some insights, so I'd setup N sets of params (separated by ===) and then in parallel run the framework to compare the results? 

Would it be possible to extend this to some form of "parameter interaction control"? A naive example: "if columnCount is even, then set wrapAround = True". 
- this could work if AE runs, and specified `Modification:` is applied to all params on each round (evaluation).

Reply:
> The other, not specified params, are the constructors defaults

Yes, if a parameter set does not explicitly modify a parameter, then its value is pulled from the "default_parameters" which are contained in the experiment python module (in this case "mnist.py").

> What is a usecase for this?

Your imagined use-case is good!  The automatic search methods are not smart.  Your human intelligence and intuition is far superior to any automatic search methods.  The search methods are dumb and rely on brute force.

Furthermore, when you try to new parameters and observe the results, you learn and you gain a deeper intuition about how the HTM works.  Manually specifying experiments lets me *play* with these things.  Also, don't forget that the complete output of the experiment is saved into the "journal" files, so you can analyze all of the data.

> parameter interaction control

The logic for this would be specific to the experiment you were running. For example the parameters for MNIST are different than the parameters for NAB, so if you wanted special logic for the MNIST parameters it would not work for NAB.  My best advice for making this would be to make a custom optimization method, like grid-search but with different custom rules.

Another option would be to make your experiment raise a `ValueError` when the invalid parameter-combination is given.  When the AE program receives certain exceptions (ValueError included) it records the error and moves on to a different set of parameters.  As long as the search method has other parameters to try which don't crash, very little time will be spent evaluating the bad parameters.

Reply:
> Another option would be to make your experiment raise a ValueError when the invalid parameter-combination is given. When the AE program receives certain exceptions (ValueError included) it records the error and moves on to a different set of parameters. As long as the search method has other parameters to try which don't crash, very little time will be spent evaluating the bad parameters.

this is a good advice! Thanks. Might be worth mentioning this approach in the docs.